[ICE]: `expected a projection`

<!--
[31mICE[0m: Rustc ./a.rs '-Zcrate-attr=feature(inherent_associated_types) -Znext-solver=globally -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/ty_kind.rs:522:67: 'expected a projection'', 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/ty_kind.rs:522:67: 'expected a projection''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Znext-solver=globally
#![feature(inherent_associated_types)]
pub struct Windows<T> {}

impl<T> Windows<fn(&())> {
    type AssocType = impl Send;
    fn ret(&self) -> Self::AssocType {
        ()
    }
}

````

original:
````rust
pub trait Iter {
    type Item<'a> where Self: 'a;

    fn next<'a>(&'a mut self) -> Option<Windows<T>>;

    fn for_each<F>(&'a mut self, mut f: F)
        where Self: Sized, F: for<'a> FnMut(Self::Windows<'a>)
    {
      todo!()
    }
}

pub struct Windows<T> {}

impl<T> Windows<fn(&())> {
    type AssocType = impl Send;
    fn ret(&self) -> Self::AssocType {
        ()
    }
}

impl<T> Iter for Windows<T> {

    fn T<'a>(&'for_each mut self) -> Option<Self::Item<'a>> {
        todo!();
    }
}

````

Version information
````
rustc 1.97.0-nightly (532f8f144 2026-04-12)
binary: rustc
commit-hash: 532f8f1443be8f605bcc9da7968e3476d3f27d26
commit-date: 2026-04-12
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.2

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/532f8f1443be8f605bcc9da7968e3476d3f27d26/compiler/rustc_type_ir/src/ty_kind.rs#L516-L528

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zcrate-attr=feature(inherent_associated_types) -Znext-solver=globally`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: `impl Trait` in associated types is unstable
 --> /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:4:22
  |
4 |     type AssocType = impl Send;
  |                      ^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(impl_trait_in_assoc_type)]` to the crate attributes to enable
  = note: this compiler was built on 2026-04-12; consider upgrading it if it is out of date

warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
 --> <crate attribute>:1:12
  |
1 | #![feature(inherent_associated_types)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:8:2
  |
8 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs`

error[E0392]: type parameter `T` is never used
 --> /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:1:20
  |
1 | pub struct Windows<T> {}
  |                    ^ unused type parameter
  |
  = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
  = help: if you intended `T` to be a const parameter, use `const T: /* Type */` instead

error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
 --> /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:3:6
  |
3 | impl<T> Windows<fn(&())> {
  |      ^ unconstrained type parameter

error[E0282]: type annotations needed
 --> /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:5:22
  |
5 |     fn ret(&self) -> Self::AssocType {
  |                      ^^^^^^^^^^^^^^^ cannot infer type

error: unconstrained opaque type
 --> /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:4:22
  |
4 |     type AssocType = impl Send;
  |                      ^^^^^^^^^
  |
  = note: `AssocType` must be used in combination with a concrete type within the same impl


thread 'rustc' (4172499) panicked at /rustc-dev/532f8f1443be8f605bcc9da7968e3476d3f27d26/compiler/rustc_type_ir/src/ty_kind.rs:522:67:
expected a projection
stack backtrace:
   0:     0x7fc5e342b31b - <<std[30bf32b9e108fde8]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[afc6995f3fa91ad]::fmt::Display>::fmt
   1:     0x7fc5e3a20a48 - core[afc6995f3fa91ad]::fmt::write
   2:     0x7fc5e34425e6 - <std[30bf32b9e108fde8]::sys::stdio::unix::Stderr as std[30bf32b9e108fde8]::io::Write>::write_fmt
   3:     0x7fc5e3401618 - std[30bf32b9e108fde8]::panicking::default_hook::{closure#0}
   4:     0x7fc5e341e983 - std[30bf32b9e108fde8]::panicking::default_hook
   5:     0x7fc5e23fb5bc - std[30bf32b9e108fde8]::panicking::update_hook::<alloc[ad9adc906e64577e]::boxed::Box<rustc_driver_impl[56c0fec0d7fc84ae]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7fc5e341ec62 - std[30bf32b9e108fde8]::panicking::panic_with_hook
   7:     0x7fc5e340170a - std[30bf32b9e108fde8]::panicking::panic_handler::{closure#0}
   8:     0x7fc5e33f5c39 - std[30bf32b9e108fde8]::sys::backtrace::__rust_end_short_backtrace::<std[30bf32b9e108fde8]::panicking::panic_handler::{closure#0}, !>
   9:     0x7fc5e340313d - __rustc[a6530efb79a2fd0]::rust_begin_unwind
  10:     0x7fc5dff5d63c - core[afc6995f3fa91ad]::panicking::panic_fmt
  11:     0x7fc5e311738e - <rustc_trait_selection[6b2e4746fb1d1412]::error_reporting::TypeErrCtxt>::suggest_constraint::<&<rustc_trait_selection[6b2e4746fb1d1412]::error_reporting::TypeErrCtxt>::expected_projection::{closure#0}>
  12:     0x7fc5e31e866e - <rustc_trait_selection[6b2e4746fb1d1412]::error_reporting::TypeErrCtxt>::note_and_explain_type_err
  13:     0x7fc5e1c1930a - <rustc_trait_selection[6b2e4746fb1d1412]::error_reporting::TypeErrCtxt>::note_type_err
  14:     0x7fc5e31d5c7f - <rustc_trait_selection[6b2e4746fb1d1412]::error_reporting::TypeErrCtxt>::report_and_explain_type_error
  15:     0x7fc5e0950f4d - <rustc_trait_selection[6b2e4746fb1d1412]::error_reporting::TypeErrCtxt>::report_mismatched_types
  16:     0x7fc5e1b12433 - <rustc_hir_typeck[4630abcebbdfa870]::coercion::CoerceMany>::report_return_mismatched_types
  17:     0x7fc5e3d5fe9d - <rustc_hir_typeck[4630abcebbdfa870]::fn_ctxt::FnCtxt>::check_expr_block
  18:     0x7fc5e3dbc5c1 - <rustc_hir_typeck[4630abcebbdfa870]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  19:     0x7fc5e3d552c0 - rustc_hir_typeck[4630abcebbdfa870]::check::check_fn
  20:     0x7fc5e4c01733 - rustc_hir_typeck[4630abcebbdfa870]::typeck_with_inspect::{closure#0}
  21:     0x7fc5e4c00130 - rustc_query_impl[c84ee57d90a0448f]::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
  22:     0x7fc5e3dee466 - rustc_query_impl[c84ee57d90a0448f]::execution::try_execute_query::<rustc_data_structures[981e95ba1a8afd5c]::vec_cache::VecCache<rustc_span[786217fdcf1835d0]::def_id::LocalDefId, rustc_middle[5ed6dbbca286abee]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[5ed6dbbca286abee]::dep_graph::graph::DepNodeIndex>, false>
  23:     0x7fc5e3dedc83 - rustc_query_impl[c84ee57d90a0448f]::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
  24:     0x7fc5e3df17d4 - <rustc_middle[5ed6dbbca286abee]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis[a31e2ccfc121be5e]::check_crate::{closure#1}>::{closure#0}
  25:     0x7fc5e3df11ad - rustc_hir_analysis[a31e2ccfc121be5e]::check_crate
  26:     0x7fc5e490d742 - rustc_interface[82dfd3b1fd905165]::passes::analysis
  27:     0x7fc5e4bbe0c9 - rustc_query_impl[c84ee57d90a0448f]::execution::try_execute_query::<rustc_middle[5ed6dbbca286abee]::query::caches::SingleCache<rustc_middle[5ed6dbbca286abee]::query::erase::ErasedData<[u8; 0usize]>>, false>
  28:     0x7fc5e4bbdd29 - rustc_query_impl[c84ee57d90a0448f]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  29:     0x7fc5e4b94890 - rustc_interface[82dfd3b1fd905165]::interface::run_compiler::<(), rustc_driver_impl[56c0fec0d7fc84ae]::run_compiler::{closure#0}>::{closure#1}
  30:     0x7fc5e4b7e8be - std[30bf32b9e108fde8]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[82dfd3b1fd905165]::util::run_in_thread_with_globals<rustc_interface[82dfd3b1fd905165]::util::run_in_thread_pool_with_globals<rustc_interface[82dfd3b1fd905165]::interface::run_compiler<(), rustc_driver_impl[56c0fec0d7fc84ae]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  31:     0x7fc5e4b7f160 - <std[30bf32b9e108fde8]::thread::lifecycle::spawn_unchecked<rustc_interface[82dfd3b1fd905165]::util::run_in_thread_with_globals<rustc_interface[82dfd3b1fd905165]::util::run_in_thread_pool_with_globals<rustc_interface[82dfd3b1fd905165]::interface::run_compiler<(), rustc_driver_impl[56c0fec0d7fc84ae]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[afc6995f3fa91ad]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  32:     0x7fc5e4b7ffec - <std[30bf32b9e108fde8]::sys::thread::unix::Thread>::new::thread_start
  33:     0x7fc5de6a597a - <unknown>
  34:     0x7fc5de7292bc - <unknown>
  35:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.97.0-nightly (532f8f144 2026-04-12) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z crate-attr=feature(inherent_associated_types) -Z next-solver=globally -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck_root] type-checking `<impl at /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:3:1: 3:25>::ret`
#1 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 6 previous errors; 1 warning emitted

Some errors have detailed explanations: E0207, E0282, E0392, E0601, E0658.
For more information about an error, try `rustc --explain E0207`.

```

</p>
</details>

<!--
query stack:
#0 [typeck_root] type-checking `<impl at /tmp/icemaker_global_tempdir.nWLI7VjApHkt/rustc_testrunner_tmpdir_reporting.cMBT7uk0gMpW/mvce.rs:3:1: 3:25>::ret`
#1 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-inherent_associated_types +WG-trait-system-refactor

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
@@ -606,6 +606,9 @@ impl<T> Trait<T> for X {
         ty: Ty<'tcx>,
     ) -> bool {
         let tcx = self.tcx;
+        if !matches!(proj_ty.kind, ty::AliasTyKind::Projection { .. }) {
+            return false;
+        }
         let Some(body_owner_def_id) = body_owner_def_id else {
             return false;
         };

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
@@ -606,6 +606,7 @@ impl<T> Trait<T> for X {
         ty: Ty<'tcx>,
     ) -> bool {
         let tcx = self.tcx;
+        // FIXME(inherent_associated_types): Extend this to support `ty::Inherent`, too.
         if !matches!(proj_ty.kind, ty::AliasTyKind::Projection { .. }) {
             return false;
         }

--- a/tests/ui/associated-inherent-types/next-solver-opaque-inherent-fn-ptr-issue-155204.rs
+++ b/tests/ui/associated-inherent-types/next-solver-opaque-inherent-fn-ptr-issue-155204.rs
@@ -1,0 +1,18 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/155204>
+//@compile-flags: -Znext-solver=globally
+#![feature(inherent_associated_types)]
+pub struct Windows<T> {}
+//~^ ERROR type parameter `T` is never used
+
+impl<T> Windows<fn(&())> {
+    //~^ ERROR the type parameter `T` is not constrained by the impl trait, self type, or predicates
+    type AssocType = impl Send;
+    //~^ ERROR `impl Trait` in associated types is unstable
+    //~| ERROR unconstrained opaque type
+    fn ret(&self) -> Self::AssocType {
+        //~^ ERROR type annotations needed
+        ()
+        //~^ ERROR mismatched types
+    }
+}
+//~^ ERROR `main` function not found

--- a/tests/ui/associated-inherent-types/next-solver-opaque-inherent-fn-ptr-issue-155204.stderr
+++ b/tests/ui/associated-inherent-types/next-solver-opaque-inherent-fn-ptr-issue-155204.stderr
@@ -1,0 +1,63 @@
+error[E0658]: `impl Trait` in associated types is unstable
+  --> $DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs:9:22
+   |
+LL |     type AssocType = impl Send;
+   |                      ^^^^^^^^^
+   |
+   = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
+   = help: add `#![feature(impl_trait_in_assoc_type)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+
+error[E0601]: `main` function not found in crate `next_solver_opaque_inherent_fn_ptr_issue_155204`
+  --> $DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs:17:2
+   |
+LL | }
+   |  ^ consider adding a `main` function to `$DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs`
+
+error[E0392]: type parameter `T` is never used
+  --> $DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs:4:20
+   |
+LL | pub struct Windows<T> {}
+   |                    ^ unused type parameter
+   |
+   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
+   = help: if you intended `T` to be a const parameter, use `const T: /* Type */` instead
+
+error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
+  --> $DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs:7:6
+   |
+LL | impl<T> Windows<fn(&())> {
+   |      ^ unconstrained type parameter
+
+error[E0282]: type annotations needed
+  --> $DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs:12:22
+   |
+LL |     fn ret(&self) -> Self::AssocType {
+   |                      ^^^^^^^^^^^^^^^ cannot infer type
+
+error: unconstrained opaque type
+  --> $DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs:9:22
+   |
+LL |     type AssocType = impl Send;
+   |                      ^^^^^^^^^
+   |
+   = note: `AssocType` must be used in combination with a concrete type within the same impl
+
+error[E0308]: mismatched types
+  --> $DIR/next-solver-opaque-inherent-fn-ptr-issue-155204.rs:14:9
+   |
+LL |     fn ret(&self) -> Self::AssocType {
+   |                      --------------- expected `Windows<for<'a> fn(&'a ())>::AssocType` because of return type
+LL |
+LL |         ()
+   |         ^^ types differ
+   |
+   = note: expected associated type `Windows<for<'a> fn(&'a ())>::AssocType`
+                    found unit type `()`
+   = help: consider constraining the associated type `Windows<for<'a> fn(&'a ())>::AssocType` to `()` or calling a method that returns `Windows<for<'a> fn(&'a ())>::AssocType`
+   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
+
+error: aborting due to 7 previous errors
+
+Some errors have detailed explanations: E0207, E0282, E0308, E0392, E0601, E0658.
+For more information about an error, try `rustc --explain E0207`.

--- a/tests/ui/associated-inherent-types/next-solver-opaque-inherent-no-ice.rs
+++ b/tests/ui/associated-inherent-types/next-solver-opaque-inherent-no-ice.rs
@@ -1,0 +1,17 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/154333>
+//@compile-flags: -Znext-solver=globally
+#![feature(inherent_associated_types)]
+
+struct Foo;
+impl<const X: y> Foo {
+    //~^ ERROR the const parameter `X` is not constrained by the impl trait, self type, or predicates
+    //~| ERROR cannot find type `y` in this scope
+    type ImplTrait = impl Clone;
+    //~^ ERROR `impl Trait` in associated types is unstable
+    //~| ERROR unconstrained opaque type
+    fn f() -> Self::ImplTrait {
+        ()
+        //~^ ERROR mismatched types
+    }
+}
+//~^ ERROR `main` function not found

--- a/tests/ui/associated-inherent-types/next-solver-opaque-inherent-no-ice.stderr
+++ b/tests/ui/associated-inherent-types/next-solver-opaque-inherent-no-ice.stderr
@@ -1,0 +1,56 @@
+error[E0425]: cannot find type `y` in this scope
+  --> $DIR/next-solver-opaque-inherent-no-ice.rs:6:15
+   |
+LL | impl<const X: y> Foo {
+   |               ^ not found in this scope
+
+error[E0658]: `impl Trait` in associated types is unstable
+  --> $DIR/next-solver-opaque-inherent-no-ice.rs:9:22
+   |
+LL |     type ImplTrait = impl Clone;
+   |                      ^^^^^^^^^^
+   |
+   = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
+   = help: add `#![feature(impl_trait_in_assoc_type)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+
+error[E0601]: `main` function not found in crate `next_solver_opaque_inherent_no_ice`
+  --> $DIR/next-solver-opaque-inherent-no-ice.rs:16:2
+   |
+LL | }
+   |  ^ consider adding a `main` function to `$DIR/next-solver-opaque-inherent-no-ice.rs`
+
+error[E0207]: the const parameter `X` is not constrained by the impl trait, self type, or predicates
+  --> $DIR/next-solver-opaque-inherent-no-ice.rs:6:6
+   |
+LL | impl<const X: y> Foo {
+   |      ^^^^^^^^^^ unconstrained const parameter
+   |
+   = note: expressions using a const parameter must map each value to a distinct output value
+   = note: proving the result of expressions other than the parameter are unique is not supported
+
+error: unconstrained opaque type
+  --> $DIR/next-solver-opaque-inherent-no-ice.rs:9:22
+   |
+LL |     type ImplTrait = impl Clone;
+   |                      ^^^^^^^^^^
+   |
+   = note: `ImplTrait` must be used in combination with a concrete type within the same impl
+
+error[E0308]: mismatched types
+  --> $DIR/next-solver-opaque-inherent-no-ice.rs:13:9
+   |
+LL |     fn f() -> Self::ImplTrait {
+   |               --------------- expected `Foo::ImplTrait` because of return type
+LL |         ()
+   |         ^^ types differ
+   |
+   = note: expected associated type `Foo::ImplTrait`
+                    found unit type `()`
+   = help: consider constraining the associated type `Foo::ImplTrait` to `()` or calling a method that returns `Foo::ImplTrait`
+   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
+
+error: aborting due to 6 previous errors
+
+Some errors have detailed explanations: E0207, E0308, E0425, E0601, E0658.
+For more information about an error, try `rustc --explain E0207`.

